core/src/com/unciv/logic/map/mapunit/MapUnit.kt

```diff
@@ -645,8 +645,21 @@ class MapUnit : IsPartOfGameInfoSerialization {
             if (tile.isEnemyTerritory(civ)) return false
             return buildImprovementUniques.any()
         }
-        return buildImprovementUniques
-                .any { improvement.matchesFilter(it.params[0], cache.state) || tile.matchesTerrainFilter(it.params[0], civ) }
+
+        // Validate that the improvement is available for the unit
+        if (!improvement.getMatchingUniques(UniqueType.OnlyAvailable, GameContext.IgnoreConditionals)
+                .none { unique -> !unique.conditionalsApply(cache.state) } ||
+            !improvement.getMatchingUniques(UniqueType.Unavailable, GameContext.IgnoreConditionals)
+                .none { unique -> unique.conditionalsApply(cache.state) }) {
+            return false
+        }
```

Comment:
Instead of "not exist none" use "exist any", more readable 
"If there exists an only available that is unsatistfied, OR there exists an unavailable that is satisfied"